### ISSUE TITLE
Batch AppendEntriesRequests and AppendEntriesResponses

### ISSUE THREAD

**Issue by belaban** [belaban/jgroups-raft#116]
When a leader sends updates 100 updates (e.g. 400-500), then rather than sending 100 `AppendEnriesRequests`, it could send a single request with 100 updates. This would cause *1* write to the log (e.g. using a `WriteBatch` in LevelDB), rather than 100 writes. It would also return a single `AppendEntriesResponse` with 100 `AppendResults` (batched).

The leader would use the 100 AppendResults to notify the 100 clients and then do a single update of `CommitTable`.
This should increase performance dramatically, as we're reducing the number of writes to the log.

**Comment by franz1981** [comment]
I see that https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf contains these nice chapters that could be used to safely applies optimizations like these, see
- 10.2.1 Writing to the leader’s disk in parallel
- 10.2.2 Batching and pipelining

**Comment by belaban** [comment]
Right, that's why I created this issue, at least for 10.2.2! Perhaps we should create a separate issue for 10.2.1

**Comment by belaban** [comment]
Overview of batching (example: 100 requests):
* Leader appends 100 entries to its (fresh) log and changes `last-appended` to 100 in *one* operation (ie., a WriteBatch in LevelDBLog)
* Leader sends a batch of 100 requests to all followers
* A follower does the necessary checks (prev entry and prev entry's term match), then appens 100 entries and sets `last-appended` in *one* operation (1 fsync)
* The followers send back an `AppendEntriesResponse` with index=100; this acks 100 and all lower indices
* The leader gets the response from follower L and adds L's vote to entry 100 in the request table and all lower-numbered requests
* When there is a majority, log entries 1-100 are applies to the state and `commit-index` is advanced to 100
* On the next request, or after `send-interval` ms, the commit of 100 is sent to the followers: each follower also applies log entries 1-100 to its state and advances `commit-index`

**Comment by belaban** [comment]
The temp drain metrics I added shows that for every request, we get N-1 (N: cluster size) responses (see the drain ratio on the leader: 33% down requests and 67% up requests (responses):
```
#1 (220 bytes):
C [ip=192.168.1.111:7802, 3 mbr(s), cluster=counter-perf, version=5.2.2.Final (Stelvio)]
RAFT={drain_ratio=down=0,00 up=1,00, drained_avg=min/avg/max=1/7,82/374, drained_down=0, drained_total=488400, drained_up=488400}

#2 (220 bytes):
B [ip=192.168.1.111:7801, 3 mbr(s), cluster=counter-perf, version=5.2.2.Final (Stelvio)]
RAFT={drain_ratio=down=0,00 up=1,00, drained_avg=min/avg/max=1/7,84/947, drained_down=0, drained_total=488401, drained_up=488401}

#3 (228 bytes):
A [ip=192.168.1.111:7800, 3 mbr(s), cluster=counter-perf, version=5.2.2.Final (Stelvio)]
RAFT={drain_ratio=down=0,33 up=0,67, drained_avg=min/avg/max=1/403,52/627, drained_down=488398, drained_total=1465199, drained_up=976801}
```

**Comment by belaban** [comment]
OK, the drain rate changed dramatically: (on the leader)
```
A [ip=192.168.1.111:7800, 3 mbr(s), cluster=counter-perf, version=5.2.2.Final (Stelvio)]
RAFT={drain_ratio=down=0,99 up=0,01, drained_avg=min/avg/max=1/93,03/299, drained_down=616999, drained_total=626302, drained_up=9303}
```

Explanation: most of the elements in the queue are now *down requests*: 100 updater threads add their update requests to the queue, and the 2 followers send only *1* ack (instead of 100).

**Comment by belaban** [comment]
Results of counter-perf (3 members, 100 threads, MacOS) on _master (`log_use_fsync="true"`):
```
B: 805,26 updates/sec (49.471 updates, 123,39ms / update)
C: 804,34 updates/sec (49.424 updates, 124,19ms / update)
A: 807,31 updates/sec (49.598 updates, 122,65ms / update)


Throughput: 805,63 updates/sec/node
Time:       123,22ms / update
```

**Comment by belaban** [comment]
Result of counter-perf (3 members, 100 threads, MacOS) on branch `raft-116-batching` (`log_use_fsync="true"`):
```
B: 8.774,87 updates/sec (526.817 updates, 11,40ms / update)
C: 8.784,11 updates/sec (527.389 updates, 11,38ms / update)
A: 9.033,33 updates/sec (542.370 updates, 11,07ms / update)


Throughput: 8.864,10 updates/sec/node
Time:       11,23ms / update
```

This is a *10x improvement*!

**Comment by belaban** [comment]
Closing; other improvements such as moving the commit index in one fsync will be tackled in other tickets
